File: cellpath.py
from __future__ import division, print_function
import numpy as np
import numba
from numba import cuda
import random
import sys
import time
import math
import pdb

width = 32768
height = 16384

def make_grid():
    arr = np.random.rand(width, height)
    arr = arr*0.9+0.1
    return arr.astype(np.float32)

def pick_start():
    return width//2, height//2
inf = float('infinity')

@cuda.jit(device=True)
def cell(grid, costs, x, y, dx, dy, epsilon):
    v1 = grid[x][y]
    c1 = inf
    c2 = inf
    c3 = inf
    c4 = inf
    if x-1 >= 0:
        c1 = grid[x-1][y]+costs[x-1][y]
    if x+1 < dx:
        c2 = grid[x+1][y]+costs[x+1][y]
    if y-1 >= 0:
        c3 = grid[x][y-1]+costs[x][y-1]
    if y+1 < dy:
        c4 = grid[x][y+1]+costs[x][y+1]
    v2 = min(v1, min(min(c1, c2), min(c3, c4)))
    writing = v1-v2 >= epsilon
    grid[x][y] = v2
    return cuda.selp(writing, 1.0, 0.0)

@cuda.jit(fastmath=True)
def kernel(grid, costs, written, dx, dy, kx, ky, epsilon, limiter):
    running = True
    writes = numba.float32(0)
    x, y = cuda.grid(2)
    while running and limiter:
        limiter -= 1
        wrote_this_time = numba.float32(0)
        for j in range(ky):
            for i in range(kx):
                wrote_this_time += cell(grid, costs, x*kx+i, y*ky+j, dx, dy, epsilon)
        writes += wrote_this_time
        runnin = wrote_this_time > 0
    cuda.atomic.add(written, (cuda.blockIdx.x, cuda.blockIdx.y), writes)

@numba.jit
def initial_pass(grid, costs, dx, dy):
    for y in range(dy):
        for x in range(1, dx):
            prev = grid[x-1][y]
            cur = grid[x][y]
            grid[x][y] = min(cur, prev+costs[x-1][y])
    for x in range(dx):
        for y in range(1, dy):
            prev = grid[x][y-1]
            cur = grid[x][y]
            grid[x][y] = min(cur, prev+costs[x][y-1])

#A workspace. Every path is infinitely long to start.
grid = np.full((width, height), float('infinity'), dtype = np.float32)
costs = make_grid()
epsilon = np.min(costs)/2.0
print(epsilon)
start = pick_start()
grid[start[0]][start[1]] = 0.0 #Going to the start is free.

stream = cuda.stream()
dgrid = cuda.to_device(grid, stream)
dcosts = cuda.to_device(costs, stream)

threadX = 16
threadY = 16
blockX = 512
blockY = 512
widthPerKernel = width//threadX//blockX
heightPerKernel = height//threadY//blockY

initial_pass(grid, costs, width, height)
written = np.zeros((blockX, blockY), dtype=np.float32)
print("Running. Start =", start)
begin = time.time()
first = True
writes = 0
writes_next = 0
limits = [1, 3, 5, 7, 9, 11, 13, 15, 17, 20, 21, 23]
lindex = 0
while writes < writes_next or first:
    first = False
    writes = writes_next
    for lim in limits:
        kernel[(blockX, blockY), (threadX, threadY), stream](dgrid, dcosts, written, width, height, widthPerKernel, heightPerKernel, epsilon, lim)
        lindex += 1
    stream.synchronize()
    writes_next = int(written.sum())
    print(writes_next-writes)
end = time.time()
print("Total time:", end-begin)
print("Wrote", writes)

File: cellpath.cpp
//This is a C++ implementation for timing comparisons.
#include <random>
#include <stdio.h>
#include <time.h>
#include <limits>

const int WIDTH = 101;
const int HEIGHT = 101;
float workspace[WIDTH][HEIGHT];
float costs[WIDTH][HEIGHT];
const int START_X = 50;
const int START_Y = 50;
#define GET_INFINITY (std::numeric_limits<float>::infinity())

inline float min2(float a, float b){
	return a < b ? a : b;
}

inline float min4(float a, float b, float c, float d) {
	float p = min2(a, b);
	float q = min2(c, d);
	return min2(p, q);
}

template<typename T1, typename T2>
int automaton(T1 workspace, T2 costs, float epsilon) {
	int writes = 0;
	const float inf = GET_INFINITY;
	for(int x = 0; x < WIDTH; x++) {
		for(int y = 0; y < HEIGHT; y++) {
			float current = workspace[x][y];
			float tmp = min4(
				x+1 < WIDTH ? workspace[x+1][y]+0.5*costs[x+1][y] : inf,
				x-1 >= 0 ? workspace[x-1][y]+0.5*costs[x-1][y] : inf,
				y+1 < HEIGHT ? workspace[x][y+1]+0.5*costs[x][y+1] : inf,
				y-1 >= 0 ? workspace[x][y-1]+0.5*costs[x][y-1] : inf);
			tmp += 0.5*costs[x][y];
			if(tmp+epsilon < current) {
				workspace[x][y] = tmp;
				writes++;
			}			
		}
	}
	return writes;
}

//returns the epsilon
float prepare() {
	const float inf = GET_INFINITY;
	float epsilon = inf;
	std::mt19937 engine{};
	std::uniform_real_distribution<> gen{0.1, 1.0};
	for(int x = 0; x < WIDTH; x++) {
		for(int y = 0; y < HEIGHT; y++) {
			costs[x][y] = gen(engine);
			epsilon  = min2(epsilon, costs[x][y]);
			workspace[x][y] = inf;
		}
	}
	workspace[START_X][START_Y] = 0.0f;
	return epsilon/2.0;
}

template<typename T1, typename T2>
int mainloop(T1 workspace, T2 costs, float epsilon) {
	int writes_total = 0;
	int writes = 0;
	do {
		writes = automaton(workspace, costs, epsilon);
		writes_total += writes;
	} while(writes);
	return writes_total;
}

void main() {
	float epsilon = prepare();
	auto start = clock();
	int writes = mainloop(workspace, costs, epsilon);
	auto end = clock();
	double secs = (float)(end-start)/CLOCKS_PER_SEC;
	printf("Total time: %f\n", secs);
	printf("Writes: %i\n", writes);
}
